### PR TITLE
Rework interop handling

### LINKED ISSUE #2554: Rollup's external default import shim is not entirely compatible with TypeScript's --esModuleInterop
When building a bundle with rollup (I have been using `experimentalCodeSplitting`), and when using an ES6 default import on a module marked as external, rollup will emit a line in the following form at the top of the chunk:

```
foo = foo && foo.hasOwnProperty('default') ? foo['default'] : foo;
```

to shim around the fact that the external may be an ES6 module that exports a default member under the `default` property, or may be a non-ES6 module that expects the entire exported value to be the "default" export.

TypeScript, as of version 2.7, has a flag `--esModuleInterop` that does a very similar thing. When importing using the import default syntax, TS emits:

```
foo = __importDefault(foo);
```

where `__importDefault` is defined in `tslib` as

```
__importDefault = function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
```

and modifies the module body to henceforth refer to `foo.default` instead of `foo`.

In addition, TS also shims around `import *` with the `tslib` helper function

```
__importStar = function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
```

Theoretically, these tslib helpers are supposed to bring TS's compiled output semantics closer to that of babel, but I have no experience with babel personally. 

### Expected Behavior / Situation
It is possible to get rollup to emit module import shims that exactly match the behavior of TS's shims.

### Actual Behavior / Situation
Rollup's default shim is subtly different from TS's `__importDefault` function. For example, if an external dependency has an exported property called `default`, but is not built with a tool that adds the `__esModule` property, rollup will treat the `default` property as the member to be imported, but TS will not.

Rollup's lack of an `import *` shim is also subtly different from TS's `__importStar` function. For example, TS makes it possible to `import * as foo from '...'` and then refer to `foo.default`, while rollup doesn't.

### Modification Proposal

There are few options here, so I'm curious to hear from the rollup team what you believe to be the best solution here. Some possible options:

- No changes to rollup upstream -- the recommended workaround is to write a `transform` hook plugin that shims `import *` lines with a shim compatible with TS's `__importStar`, and to write a `renderChunk` hook plugin that shims import default lines with a shim compatible with TS's `__importDefault`. (The former must be a `transform` hook since it is difficult to determine which lines in a compiled bundle were originally `import *`s. The latter must be a `renderChunk` hook since rollup emits its own default import shim in compiled bundles that must be removed by the plugin.) I've already succeeded at writing a functional implementation of the latter, but it seems like a messy solution overall, especially if we expect others to also encounter this situation.

- We add a new compile option/flag to rollup to optionally make it emit tslib-compatible shims for import default and `import *` instead of the current behavior. This doesn't break backwards compatibility for existing users, and is a clean solution for anyone else who might need compatibility with TS's `esModuleInterop` in the future. However, this does involve a code change to rollup hard-coding in a use-case-specific implementation detail.

- We add a config option that allows one to specify the ES import shim implementation that should be emitted in compiled bundles. This means the default implementation could be overriden by the user, and it would be the user's responsibility to pass something compatible with tslib.

- Alternately, if adding a new config option with a template or code-generation pattern is deemed too heavyweight of an interface, there could be a new boolean config option that merely disables the existing default import shim, where one would need to continue to rely on rollup plugins to codemod TS-compatible shims back into the compiled bundles.

Look forward to hearing your thoughts!

### LINKED ISSUE #3288: interopNamespace added when interop is disabled
- Rollup Version: 1.27.9
- Operating System (or Browser): Windows 10
- Node Version: 10.15.3

### How Do We Reproduce?

```js
// rollup.config.js
config.output.format = 'cjs';
config.output.interop = false;

// entry
import('someExternal');
```
### Expected Behavior
```js
new Promise(function (resolve) { resolve(require('someExternal')); });
```

### Actual Behavior
```js
new Promise(function (resolve) { resolve(_interopNamespace(require('someExternal'))); });
```

### PR DESCRIPTION
**Pull request by lukastaegert**
This PR contains:
- [x] bugfix
- [x] feature
- [ ] refactor
- [x] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:
Resolves #3288 
Resolves #2554 

## Description
This completely reimplements how interop is handled. I.e. in formats cjs, amd, iife and umd, this reimplements how default and namespace imports and reexports are handled, as well as what is returned by dynamic imports from external dependencies.

Besides the existing values, there are now four new interop values: "auto", "esModule", "default" and "defaultOnly". Moreover, a function can be supplied for `output.interop` to configure the correct interop per dependency.

From the new documentation:

#### output.interop
Type: `"auto" | "esModule" | "default" | "defaultOnly" | boolean | ((id: string) => "auto" | "esModule" | "default" | "defaultOnly" | boolean)`<br>
CLI: `--interop <value>`<br>
Default: `true`

Controls how Rollup handles default, namespace and dynamic imports from external dependencies in formats like CommonJS that do not natively support these concepts. Note that even though `true` is the current default value, this value is deprecated and will be replaced by `"auto"` in the next major version of Rollup. In the examples, we will be using the CommonJS format, but the interop similarly applies to AMD, IIFE and UMD targets as well.

To understand the different values, assume we are bundling the following code for a `cjs` target:

```js
import ext_default, * as external from 'external1';
console.log(ext_default, external.bar, external);
import('external2').then(console.log);
```

Keep in mind that for Rollup, `import * as ext_namespace from 'external'; console.log(ext_namespace.bar);` is completely equivalent to `import {bar} from 'external'; console.log(bar);` and will produce the same code. In the example above however, the namespace object itself is passed to a global function as well, which means we need it as a properly formed object.

- `"esModule"` assumes that required modules are transpiled ES modules where the required value corresponds to the module namespace and the default export is the `.default` property of the exported object:

  ```js
  var external = require('external1');
  console.log(external['default'], external.bar, external);
  Promise.resolve().then(function () {
    return require('external2');
  }).then(console.log);
  ```
  
  When `esModule` is used, Rollup adds no additional interop helpers and also supports live-bindings for default exports.
  
- `"default"` assumes that the required value should be treated as the default export of the imported module, just like when importing CommonJS from an ES module context in Node. In contrast to Node, though, named imports are supported as well which are treated as properties of the default import. To create the namespace object, Rollup injects helpers:
  
  ```js
  var external = require('external1');
  
  function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
      Object.keys(e).forEach(function (k) {
        if (k !== 'default') {
          var d = Object.getOwnPropertyDescriptor(e, k);
          Object.defineProperty(n, k, d.get ? d : {
            enumerable: true,
            get: function () {
              return e[k];
            }
          });
        }
      });
    }
    n['default'] = e;
    return Object.freeze(n);
  }
  
  var external__namespace = /*#__PURE__*/_interopNamespaceDefault(external);
  console.log(external, external.bar, external__namespace);
  Promise.resolve().then(function () {
    return /*#__PURE__*/_interopNamespaceDefault(require('external2'));
  }).then(console.log);
  ```
  
- `"auto"` combines both `"esModule"` and `"default"` by injecting helpers that contain code that detects at runtime if the required value contains the [`__esModule` property](guide/en/#outputesmodule). Adding this property is a standard implemented by Rollup, Babel and many other tools to signify that the required value is the namespace of a transpiled ES module:

  ```js
  var external = require('external1');
  
  function _interopNamespace(e) {
    if (e && e.__esModule) { return e; } else {
      var n = Object.create(null);
      if (e) {
        Object.keys(e).forEach(function (k) {
          if (k !== 'default') {
            var d = Object.getOwnPropertyDescriptor(e, k);
            Object.defineProperty(n, k, d.get ? d : {
              enumerable: true,
              get: function () {
                return e[k];
              }
            });
          }
        });
      }
      n['default'] = e;
      return Object.freeze(n);
    }
  }
  
  var external__namespace = /*#__PURE__*/_interopNamespace(external);
  console.log(external__namespace['default'], external.bar, external__namespace);
  Promise.resolve().then(function () {
    return /*#__PURE__*/_interopNamespace(require('external2'));
  }).then(console.log);
  ```
  
  Note how Rollup is reusing the created namespace object to get the `default` export. If the namespace object is not needed, Rollup will use a simpler helper:
  
  ```js
  // input
  import ext_default from 'external';
  console.log(ext_default);
  
  // output
  var ext_default = require('external');
  
  function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }
  
  var ext_default__default = /*#__PURE__*/_interopDefault(ext_default);
  console.log(ext_default__default['default']);
  ```

- `"defaultOnly"` is similar to `"default"` except for the following:
  * Named imports are forbidden. If such an import is encountered, Rollup throws an error even in `es` and `system` formats. That way it is ensure that the `es` version of the code is able to import non-builtin CommonJS modules in Node correctly.
  * While namespace reexports `export * from 'external';` are not prohibited, they are ignored and will cause Rollup to display a warning because they would not have an effect if there are no named exports.
  * When a namespace object is generated, Rollup uses a much simpler helper.
  
  Here is what Rollup will create from the example code.  Note that we removed `external.bar` from the code as otherwise, Rollup would have thrown an error because, as stated before, this is equivalent to a named import.
  
  ```js
  var ext_default = require('external1');
  
  function _interopNamespaceDefaultOnly(e) {
    return Object.freeze({__proto__: null, 'default': e});
  }
  
  var ext_default__namespace = /*#__PURE__*/_interopNamespaceDefaultOnly(ext_default);
  console.log(ext_default, ext_default__namespace);
  Promise.resolve().then(function () {
    return /*#__PURE__*/_interopNamespaceDefaultOnly(require('external2'));
  }).then(console.log);
  ```
  
- When a function is supplied, Rollup will pass each external id to this function once to control the interop type per dependency.

  As an example if all dependencies are CommonJs, the following config will ensure that named imports are only permitted from Node builtins:

  ```js
  // rollup.config.js
  import builtins from 'builtins';
  const nodeBuiltins = new Set(builtins());
  
  export default {
    // ...
    output: {
      // ...
      interop(id) {
        if (nodeBuiltins.has(id)) {
          return 'default';
        }
        return 'defaultOnly';
      }
    }
  };
  ```
  
- `true` is equivalent to `"auto"` except that it uses a slightly different helper for the default export that checks for the presence of a `default` property instead of the `__esModule` property.
  
  ☢️ *This value is deprecated and will be removed in a future Rollup version.*
  
- `false` is equivalent to using `default` when importing a default export and `esModule` when importing a namespace.
  
  ☢️ *This value is deprecated and will be removed in a future Rollup version.*

There are some additional options that have an effect on the generated interop code:

- Setting [`output.exernalLiveBindings`](guide/en/#outputexternallivebindings) to `false` will generate simplified namespace helpers as well as simplified code for extracted default imports.
- Setting [`output.freeze`](guide/en/#outputfreeze) to `false` will prevent generated interop namespace objects from being frozen.

### Features
* Allow defining interop per dependency via a function (#3710)
* Support interop "auto" as a more compatible version of "true" (#3710)
* Support interop "default" and "esModule" to avoid unnecessary interop helpers (#3710)
* Support interop "defaultOnly" for simplified helpers and Node ESM interop compatible output (#3710)
* Respect interop option for external dynamic imports (#3710)
* Support live-bindings for external default imports in non-ES formats unless "externalLiveBindings" is "false" (#3710)
* Use shared default interop helpers for AMD, UMD and IIFE formats (#3710)
* Avoid unnecessarily deconflicted module variables in non-ES formats (#3710)
* Freeze generated interop namespace objects (#3710)
* Always mark interop helpers as pure (#3710)
* Avoid default export interop if there is already an interop namespace object (#3710)
* Sort all `require` statements to the top in CommonJS output for easier back-transpilation to ES modules by other tools (#3710)

### Bug Fixes
* Deconflict the names of helper variables introduced for interop (#3710)
* Generate proper namespace objects for static namespace imports in non-ES formats (#3710)
* Do not add unused interop helpers when using the renderDynamicImport hook (#3710)

### CODE REVIEW

File: test/form/samples/deprecated/interop-false-reexport/_expected/amd.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ define(['exports', 'external'], function (exports, external) { 'use strict';
 
 
 
-	exports.p = external.default;
+	exports.p = external;
```
Comment (guybedford) [inline test/form/samples/deprecated/interop-false-reexport/_expected/amd.js:5]: Is this a bug fix!?

Comment (lukastaegert) [reply test/form/samples/deprecated/interop-false-reexport/_expected/amd.js:5]: As a matter of fact yes, here, reexports behaved differently from regular imports where the `.default` was not added.

**Comment by lukastaegert** [comment]
@guybedford I added the documentation and also pasted it into the PR description. By tomorrow, I will try to compile what other changes are included as some hidden bugs were fixed as well.

**Comment by lukastaegert** [comment]
I also compiled the changelog now, this is a list of the included fixes and improvements:

### Features
* Allow defining interop per dependency via a function (#3710)
* Support interop "auto" as a more compatible version of "true" (#3710)
* Support interop "default" and "esModule" to avoid unnecessary interop helpers (#3710)
* Support interop "defaultOnly" for simplified helpers and Node ESM interop compatible output (#3710)
* Support live-bindings for external default imports in non-ES formats unless "externalLiveBindings" is "false" (#3710)
* Use shared default interop helpers for AMD, UMD and IIFE formats (#3710)
* Avoid unnecessarily deconflicted module variables in non-ES formats (#3710)
* Freeze generated interop namespace objects (#3710)
* Always mark interop helpers as pure (#3710)
* Avoid default export interop if there is already an interop namespace object (#3710)

### Bug Fixes
* Deconflict the names of helper variables introduced for interop (#3710)
* Generate proper namespace objects for static namespace imports in non-ES formats (#3710)
* Do not add unused interop helpers when using the renderDynamicImport hook (#3710)

**Comment by lukastaegert** [comment]
This is now complete and ready for review
cc @guybedford

**Comment by lukastaegert** [comment]
This is a tricky one. The best I could think of is
Detect if by removing an `else` branch, another `else` branch can be caught (i.e. the rendered parent is an if statement and we are in the consequent branch and there exists an alternate branch) and if that is the case, do not remove the branch but replace it with an empty `else {}` branch.

**Comment by dgoldstein0** [comment]
very impressive list of changes!  Definitely looking forward to this.

The description mentions that dynamic imports are impacted too, but doesn't give any examples of the codegen changes for those.  could you add more details around that?  I expect it'll be the same as the `import *` changes, but would like some confirmation.

**Comment by lukastaegert** [comment]
Yes, exactly. Maybe I extend the examples to also contain a dynamic import as it would of course use the same helper.

**Comment by guybedford** [comment]
Also, the modern module brand check can be done with `Object.prototype.toString.call(ns) === '[object Module]'` to support legacy environments properly.

**Comment by lukastaegert** [comment]
@guybedford That is definitely an interesting idea. If I understand correctly, though, it would involve Rollup's cjs/amd/iife/umd output to replace or augment

```js
Object.defineProperty(exports, '__esModule', { value: true });
```

with

```js
exports[Symbol.toStringTag] = 'Module';
```

or alternatively

```js
Object.defineProperty(exports, 'toString', { value: function() { return  '[object Module]'; } });
```

for a legacy version to work correctly?

@dgoldstein0 I extended the examples to also contain a dynamic import as well, I hope this answers your question.

**Comment by guybedford** [comment]
@lukastaegert actually thinking about this more I'm not sure it solves anything... sorry for false suggestion. I don't think we should ever change the __esModule brand because that is a reliable differentiator. Rather this was about environment inference, but in an ES module environment we just want to be treating everything as a namespace anyway so perhaps best not to meddle.

**Comment by dgoldstein0** [comment]
> @dgoldstein0 I extended the examples to also contain a dynamic import as well, I hope this answers your question.

yes it does, thanks a ton!

**Pull request merged**